### PR TITLE
gcoap: DEVELHELP checks resources are sorted

### PR DESCRIPTION
**Pull request by cladmi**
gcoap requires resources to be sorted alphabetically, but currently nothing verifies it.
I added a runtime check that warns if not the case, only enable when DEVELHELP is.

Also, in tests-gcoap, the test resources where not sorted.

To make it visible, without adding many tests, I needed to enable DEVELHELP in unittests makefile.

I also fixed an unused variable enclosing ifdef, as its NDEBUG that controls assert.
NDEBUG should be enabled when DEVELHELP is, but if you enable it with -DDEVELHELP=1, NDEBUG is not and it did not compile anymore.

### CODE REVIEW

**Comment by aabadie** [comment]
Thanks @cladmi and welcome to RIOT :)

File: sys/net/application_layer/gcoap/gcoap.c [hunk 1]
```diff
@@ -570,6 +570,32 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
 }
 
 /*
+ * Check listener structure
+ *
+ *  - resources should be sorted alphabetically
+ */
+#ifdef DEVELHELP
+static void _check_listener(gcoap_listener_t *listener)
```
Comment (smlng) [inline sys/net/application_layer/gcoap/gcoap.c]: IMHO a check function should have return value, i.e., `0` on success and anything else on error.

File: sys/net/application_layer/gcoap/gcoap.c [hunk 2]
```diff
@@ -570,6 +570,32 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
...
+{
+    const char *prev_path = "";
+    const char *path = NULL;
+
+    for (size_t i = 0; i < listener->resources_len; i++) {
```
Comment (smlng) [inline sys/net/application_layer/gcoap/gcoap.c]: Why not start with `i = 1` and use `listener->resources[i-1]->path` and `listener->resources[i]->path` in `strcmp` that way omitting these extra variables plus remembering the previous path.

File: sys/net/application_layer/gcoap/gcoap.c [hunk 3]
```diff
@@ -594,6 +620,8 @@ kernel_pid_t gcoap_init(void)
 
 void gcoap_register_listener(gcoap_listener_t *listener)
 {
+    _check_listener(listener);
```
Comment (smlng) [inline sys/net/application_layer/gcoap/gcoap.c]: if (as suggested above) check_listener would have a return value, you could use `assert(!_check_listener(listener);` here.

File: sys/net/application_layer/gcoap/gcoap.c [hunk 4]
```diff
@@ -806,7 +834,7 @@ uint8_t gcoap_op_state(void)
 int gcoap_get_resource_list(void *buf, size_t maxlen, uint8_t cf)
 {
     assert(cf == COAP_CT_LINK_FORMAT);
-#ifndef DEVELHELP
+#ifdef NDEBUG
```
Comment (smlng) [inline sys/net/application_layer/gcoap/gcoap.c]: why this change?

Comment (cladmi) [reply sys/net/application_layer/gcoap/gcoap.c]: The definition of `assert` depends on the definition of `NDEBUG`
`NDEBUG` is defined by the build system if `DEVELHELP`is defined with `-DDEVELHELP`.
https://github.com/RIOT-OS/RIOT/blob/2017.07/makefiles/cflags.inc.mk#L55

So this `ifdef` worked because of the build system, and if you do `CFLAGS += -DDEVELHELP=1`, `assert` is not defined and you get an unused variable warning.

It may belong in another PR though.

**Comment by cladmi** [comment]
This check should be pushed to `nanocoap` directly as it is a `nanocoap` requirement.
I will do it when `nanocoap` is integrated into RIOT.

**Comment by aabadie** [comment]
> I will do it when nanocoap is integrated into RIOT.

When this will happen ? (I think this is a good news personally)

**Comment by cladmi** [comment]
It was said by @kaspar030 during the summit, and if I understood it is still in his plans.
